File: src/main/kotlin/intellijplugin/java/siyeh_ig/psiutils/TypeUtils.java
package intellijplugin.java.siyeh_ig.psiutils;

import com.intellij.psi.*;
import gnu.trove.THashMap;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.Field;
import java.util.Map;

@SuppressWarnings({"UnusedDeclaration"})
public class TypeUtils {

    public static final Map<PsiType, String> PRIMITIVE_TYPES;

    static {
        PRIMITIVE_TYPES = new THashMap<PsiType, String>(10);
        PRIMITIVE_TYPES.put(getPsiType("BYTE"),    "B");
        PRIMITIVE_TYPES.put(getPsiType("CHAR"),    "C");
        PRIMITIVE_TYPES.put(getPsiType("DOUBLE"),  "D");
        PRIMITIVE_TYPES.put(getPsiType("FLOAT"),   "F");
        PRIMITIVE_TYPES.put(getPsiType("INT"),     "I");
        PRIMITIVE_TYPES.put(getPsiType("LONG"),    "J");
        PRIMITIVE_TYPES.put(getPsiType("SHORT"),   "S");
        PRIMITIVE_TYPES.put(getPsiType("VOID"),    "V");
        PRIMITIVE_TYPES.put(getPsiType("BOOLEAN"), "Z");
    }

    private TypeUtils() {
        // Nothing to do
    }

    public static boolean expressionHasType(@Nullable String typeName, @Nullable PsiExpression expression) {
	    return (expression != null && typeEquals(typeName, expression.getType()));
    }

    public static boolean typeEquals(@Nullable String typeName, @Nullable PsiType targetType) {
	    return (targetType != null && targetType.equalsToText(typeName));
    }

    public static boolean isJavaLangObject(@Nullable PsiType targetType) {
        return typeEquals("java.lang.Object", targetType);
    }

    public static boolean isJavaLangString(@Nullable PsiType targetType) {
        return typeEquals("java.lang.String", targetType);
    }

    public static boolean expressionHasTypeOrSubtype(@Nullable String typeName, @Nullable PsiExpression expression) {
        if (expression == null || typeName == null) {
            return false;
        }
        final PsiType type = expression.getType();

	    if (type == null || !(type instanceof PsiClassType)) {
		    return false;
	    }

		final PsiClass psiClass = ((PsiClassType) type).resolve();

	    return (psiClass != null && ClassUtils.isSubclass(psiClass, typeName));
    }

    /**
     * LONG, INT, SHORT, ... constants are moved from PsiType to PsiPrimitiveType
     * between IDEA 9 and IDEA 10. This method is used to encapsulate their access.
     * @param typeName the name of the constant field (e.g. "LONG")
     * @return the integral type instance
     */
    public static PsiType getPsiType(String typeName) {
        Field typeField = null;
        try {
            typeField = PsiPrimitiveType.class.getField(typeName);
        } catch (NoSuchFieldException e) { /* Ignore */ }

        if (typeField == null) {
            try {
                typeField = PsiType.class.getField(typeName);
            } catch (NoSuchFieldException e) { /* Ignore */ }
        }

        if (typeField != null) {
            try {
                return (PsiType) typeField.get(null);
            } catch (IllegalAccessException e) { /* Ignore */ }
        }

        return null;
    }
}

File: src/main/kotlin/intellijplugin/util/TypeUtils.kt
package intellijplugin.util

import com.intellij.psi.PsiClassType
import com.intellij.psi.PsiExpression
import com.intellij.psi.PsiPrimitiveType
import com.intellij.psi.PsiType
import gnu.trove.THashMap
import java.lang.reflect.Field

object TypeUtils {
    var PRIMITIVE_TYPES: MutableMap<PsiType?, String>? = null

    fun expressionHasType(typeName: String?, expression: PsiExpression?): Boolean {
        return expression != null && typeEquals(typeName, expression.type)
    }

    fun typeEquals(typeName: String?, targetType: PsiType?): Boolean {
        return targetType != null && targetType.equalsToText(typeName!!)
    }

    fun isJavaLangObject(targetType: PsiType?): Boolean {
        return typeEquals("java.lang.Object", targetType)
    }

    fun isJavaLangString(targetType: PsiType?): Boolean {
        return typeEquals("java.lang.String", targetType)
    }

    fun expressionHasTypeOrSubtype(typeName: String?, expression: PsiExpression?): Boolean {
        if (expression == null || typeName == null) {
            return false
        }
        val type = expression.type
        if (type == null || type !is PsiClassType) {
            return false
        }
        val psiClass = type.resolve()
        return psiClass != null && ClassUtils.isSubclass(psiClass, typeName)
    }

    /**
     * LONG, INT, SHORT, ... constants are moved from PsiType to PsiPrimitiveType
     * between IDEA 9 and IDEA 10. This method is used to encapsulate their access.
     * @param typeName the name of the constant field (e.g. "LONG")
     * @return the integral type instance
     */
    fun getPsiType(typeName: String?): PsiType? {
        var typeField: Field? = null
        try {
            typeField = PsiPrimitiveType::class.java.getField(typeName)
        } catch (e: NoSuchFieldException) { /* Ignore */
        }
        if (typeField == null) {
            try {
                typeField = PsiType::class.java.getField(typeName)
            } catch (e: NoSuchFieldException) { /* Ignore */
            }
        }
        if (typeField != null) {
            try {
                return typeField[null] as PsiType
            } catch (e: IllegalAccessException) { /* Ignore */
            }
        }
        return null
    }

    init {
        PRIMITIVE_TYPES = THashMap(10)
        PRIMITIVE_TYPES!![getPsiType("BYTE")] = "B"
        PRIMITIVE_TYPES!![getPsiType("CHAR")] = "C"
        PRIMITIVE_TYPES!![getPsiType("DOUBLE")] = "D"
        PRIMITIVE_TYPES!![getPsiType("FLOAT")] = "F"
        PRIMITIVE_TYPES!![getPsiType("INT")] = "I"
        PRIMITIVE_TYPES!![getPsiType("LONG")] = "J"
        PRIMITIVE_TYPES!![getPsiType("SHORT")] = "S"
        PRIMITIVE_TYPES!![getPsiType("VOID")] = "V"
        PRIMITIVE_TYPES!![getPsiType("BOOLEAN")] = "Z"
    }
}